### ISSUE TITLE
Index added by inheritance dropped during upgrade

### ISSUE THREAD

**Issue by guewen** [odoo/odoo#24958]
**Impacted versions:**

Found in 9.0, reproduced in 9.0 and 11.0, I assume it happens on 10.0 as well.

**Steps to reproduce:**

The situation happens when you have a base addon declaring a Many2one field without index, another addon which adds an index to the field by extending it, then a third addon that extend the same model (no matter what it does). The second and third addon do not have any dependency between them.

Base addon:
```
class ResPartner(models.Model):
    _name = 'res.partner'

    country_id = fields.Many2one('res.country')
```
Addon adding the index, depends on 'base'
```
class ResPartner(models.Model):
    _inherit = 'res.partner'

    country_id = fields.Many2one(index=True)
```
Addon doing something else on res.partner
```
class ResPartner(models.Model):
    _inherit = 'res.partner'

    my_field = fields.Char()
```

Once all 3 addons are installed, there should be an index on `res_partner.country_id`. Now if an upgrade of the third addon is done, the index is dropped.

It can be reproduced with the code from this repository: https://github.com/guewen/odoo-issues-reproduct/tree/issue-index-drop-on-update

1. Install the addons

```
odoo --workers=0 --stop-after-init -i addon1,addon2
```

2. Check presence of the index

```
$ psql
# \d res_partner
Indexes:
    ...
    "res_partner_country_id_index" btree (country_id)
```
3. Upgrade `addon1`

```
odoo --workers=0 --stop-after-init -u addon1
```

4. Check again the presence of the index

**Current behavior:**

The index is lost when an upgrade of the addon is done, as soon as there is an inherit of the same model and no dependency between the addons (addon1 and addon2).
It is because during the upgrade, when it reaches [this line](https://github.com/odoo/odoo/blob/12b11c1e6fe179abf988b1ff8c812fad08295215/odoo/fields.py#L900), the field is not yet flagged as `index=True`, because the other addon is not loaded yet. It probably depends of the order of loading then.

This is a dangerous behavior because it can go unnoticed. You add `index=True` in some addon, which get added properly, and suddenly removed when, later, you update (or install I guess) another addon.

**Expected behavior:**

The index is not dropped.

**Comment by pedrobaeza** [comment]
@rco-odoo for your interest

**Comment by Elkasitu** [comment]
Redefining the same field of the same model in different modules is **not** supported in versions prior to commit bbfb098, this doesn't mean that it can't work, it just means that it can have undefined behavior like this, it used to cause a lot of installation/uninstallation issues as explained by the commit message.

Also, can you find an instance within standard odoo code where we extend a field and add an index to it? If not, it might not even be supported...

I'll discuss this with @rco-odoo

**Comment by gdgellatly** [comment]
@guewen this is interesting.  Don't suppose you tested with other db centric fields like required to see if the not null constraints got dropped as well.

**Comment by guewen** [comment]
Good call @gdgellatly. Tested with `required=True`, same story, the `not null` constraint is added by an addon and removed when the second one is updated.

**Comment by Elkasitu** [comment]
The fix will be backported to 10.0 after a while, not sure about 9.0 yet as the module loading API is different.
